market: implement MarketTunnel

chappjc opened the pull request:
This implements two pieces of the `MarketTunnel` `interface`:

- [x] `MidGap`
- [x] `CoinLocked`

chappjc commented on server/market/market.go:
```diff
@@ -144,23 +153,42 @@ func NewMarket(ctx context.Context, mktInfo *dex.MarketInfo, storage db.DEXArchi
+		auth:               authMgr,
+		storage:            storage,
+		lockedCoins:        lockedCoins,
+		lockedCoinsByOrder: lockedCoinsByOrder,
```
In case anyone is snooping on these commits (but don't bother), note that this will have to happen at a higher level above the market so that locked coins will be tracked for all markets.

buck54321 commented on server/market/market.go:
```diff
@@ -349,26 +448,74 @@ func (m *Market) processOrder(rec *orderRecord, epoch *EpochQueue, errChan chan<
+func (m *Market) unlockCancelOrderCoins(co *order.CancelOrder) {
+	oid := co.TargetOrderID
+	// The OrderID does not indicate the canceled order's side (or asset), try
+	// the ID in both base and quote coin lockers.
```
You can get the side and assets from the `CancelOrder`, which has a `Prefix`. The data in the `Prefix` was validated by the `OrderRouter` and by the `MarketTunnel`'s `Cancelable` method, once implemented.

chappjc replied:
This was supposed to have been removed.  It was for a quick test.  However, for a CancelOrder, the Prefix doesn't indicate which side the TargetOrderID is on.  Turns out this was not needed though.  The `Matcher` deals with cancel orders and all we care about was the `unbooked` slice returned by `Match`.

buck54321 replied:
Oh yeah. Forgot the prefix doesn't have the side.

chappjc commented on dex/order/order.go:
```diff
@@ -153,23 +153,41 @@ type Order interface {
+	// TODO!!! The orders must be updated to track the current backing coins,
+	// not just the original backing coins. This is critical for partially
+	// filled orders where each fill creates change that must then be tracked as
+	// the new backing coins.
```
This is a big TODO.  We discussed this in matrix recently, but in summary, the Order needs to know the coins that are currently used to back the order.  This can change because of partial fills.

Say an order is partially filled and it stays on the books.  Now say the swap succeeds and the backing coins are spent.   From this swap that partially filled the order, there was change.  That change would need to be tracked now as it would be used to back the order with the new/reduced remaining amount.

buck54321 replied:
I don't think we should try to track the change output with the coin locker. The changes in #91 left a `TxID` method to hopefully deal with change outputs. See the [use of `(MarketTunnel).TxMonitored` by the `OrderRouter`](https://github.com/decred/dcrdex/blob/b64658ef0d9fdcc9222182e94519d643e9195fee/server/market/orderrouter.go#L520). I was guessing that the "dex-monitored" change output can be checked by asking the `Swapper` whether it is currently tracking the transaction ID. The `Swapper` now [saves the `Coins`](https://github.com/decred/dcrdex/blob/b64658ef0d9fdcc9222182e94519d643e9195fee/server/swap/swap.go#L63-L68) involved in the swap, so it should be straightforward to implement.

Implementation of coin tracking through the locker would necessitate the user submitting the change coin information with the `init` and `redeem` route payloads, and probably some messy edge cases. Notably, right now, the exchange wallet interface from #72 and #92 don't even give the user information about the coin output. It's tracked internally.

chappjc replied:
So it seems the swapStatus.coins needs to be used in the Swapper's coin locker?  I think it will be clear to you what needs to be done there when you get a look at the swap.go changes.

chappjc replied:
As it stands, I think this PR is going to fail to lock or double lock partial filled orders.

chappjc replied:
> The changes in #91 left a `TxID` method to hopefully deal with change outputs. See the [use of `(MarketTunnel).TxMonitored` by the `OrderRouter`](https://github.com/decred/dcrdex/blob/b64658ef0d9fdcc9222182e94519d643e9195fee/server/market/orderrouter.go#L520). I was guessing that the "dex-monitored" change output can be checked by asking the `Swapper` whether it is currently tracking the transaction ID. The `Swapper` now [saves the `Coins`](https://github.com/decred/dcrdex/blob/b64658ef0d9fdcc9222182e94519d643e9195fee/server/swap/swap.go#L63-L68) involved in the swap, so it should be straightforward to implement.

Is the following diff what you had in mind? https://github.com/decred/dcrdex/pull/95/files#diff-5f1d20749025f318cd1a8ba0915ebd37R418-R456

chappjc commented on server/coinlock/coinlocker.go:
```diff
@@ -0,0 +1,222 @@
+func (ac *AssetCoinLocker) LockCoins(orderCoins map[order.OrderID][]CoinID) {
+	ac.coinMtx.Lock()
+	for oid, coins := range orderCoins {
+		ac.lockedCoinsByOrder[oid] = coins
```
Note that this would need to be an append if the set of coins tracked by a given order changes.

chappjc commented on server/coinlock/coinlocker.go:
```diff
@@ -0,0 +1,222 @@
+}
+
+// UnlockOrderCoins unlocks any coins backing order.
+func (ac *AssetCoinLocker) UnlockOrderCoins(oid order.OrderID) {
```
Unlocking by order ID may prove to be in appropriate for the same reasons mentioned before.  Unlocking by CoinID may be needed.

buck54321 replied:
May prove to be a nice feature, but not a priority, IMO. Locking only the coins that have been spent is possible with no changes to the current backend interface though.

chappjc commented on server/coinlock/coinlocker.go:
```diff
@@ -0,0 +1,222 @@
+	}
+
+	return locker.OrderCoinsLocked(oid)
+}
```
These may or may not be used by the highest level DEX component, hence not commented, but kept here to facilitate the next steps.

chappjc commented on server/matcher/match.go:
```diff
@@ -135,29 +137,40 @@ func (m *Matcher) Match(book Booker, queue []order.Order) (matches []*order.Matc
 			}
 
+			// Either matched or standing unmatched => passed.
+			passed = append(passed, q)
```
It was a bug that passed only applies to matched orders, but not newly-booked/standing unbooked orders.

chappjc replied:
I haven't had to use it.  Will remove.

chappjc replied:
Actually, going to remove in https://github.com/decred/dcrdex/pull/95 since it touches Matcher too.

chappjc commented on dex/order/order.go:
```diff
@@ -153,23 +153,41 @@ type Order interface {
+func (c CoinID) String() string {
+	return hex.EncodeToString(c)
+}
+
```
I considered making a function like the following, but it would have been just for the coinlocker internals:

```go
func (c CoinID) MapKey() string {
	return string(c)
}
```

chappjc commented:
There is the [massive TODO](https://github.com/decred/dcrdex/pull/85#discussion_r353221604) regarding the potentially changing CoinID set for an order as it is partially filled, but I think that is slightly beyond the scope of this PR.

The main thing in this PR is the coinlocker types that provide a shared underlying coin locking system between book and swapper (and markets).

buck54321 commented on server/market/market.go:
```diff
@@ -327,6 +425,20 @@ func (m *Market) processOrder(rec *orderRecord, epoch *EpochQueue, errChan chan<
 	}
 
+	// Ensure that the received order does not use locked coins.
+	lockedCoins := m.coinsLocked(ord)
```
These coins are being checked in both the order router and the market. Is that the intention?

chappjc replied:
Since the epoch processing loop in `Market.runEpochs` runs concurrently with the order router handlers registered with the `AuthManager`, and coins aren't locked until `runEpochs` does `Market.processOrder`, there needs to be a check somewhere in the synchronous processing pipeline of `runEpochs`/`processOrder`.  I think we could eliminate `(*OrderRouter).checkPrefixTrade`, but the early extra check seemed fine.

buck54321 commented on server/market/orderrouter.go:
```diff
@@ -37,25 +37,25 @@ type MarketTunnel interface {
+	// DRAFT NOTE: This function could also potentially be handled by persistent
 	// storage, since active orders and active matches are tracked there.
-	OutpointLocked([]byte) bool
+	CoinLocked(coinID order.CoinID, assetID uint32) bool
```
You could also pass the `DEXCoinLocker` directly to the order book router, if that makes things easier.

chappjc replied:
OK either way.  But see what you make of my comment regarding the extra lock check in https://github.com/decred/dcrdex/pull/85#discussion_r354528345

buck54321 commented on server/market/market.go:
```diff
@@ -128,32 +135,64 @@ func (m *Market) SubmitOrderAsync(rec *orderRecord) <-chan error {
+// CoinLocked checks if a coin is locked. The asset is specified since we should
+// not assume that a CoinID for one asset cannot be made to match another
+// asset's CoinID.
+func (m *Market) CoinLocked(asset uint32, coin coinlock.CoinID) bool {
```
I made a comment in the order router too, but if it's easier to pass the `DEXCoinLocker` to the order router, this method could be dropped from `MarketTunnel`.

chappjc commented on server/book/book.go:
```diff
@@ -104,26 +131,40 @@ func (b *Book) Remove(oid order.OrderID) (*order.LimitOrder, bool) {
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+	count := b.sells.Count()
+	return b.sellOrdersN(count)
```
For sure.  Dunno why I was doing it this way.

chappjc commented:
A `book` bugfix is in progress.  Do not re-review yet.

chappjc commented:
Bug fix in https://github.com/decred/dcrdex/pull/85 is required to be merged first.

chappjc commented:
PR #97 with the book bug fix is merged and this is rebased.  Tests working now.  Also added the tests for `ActiveOrderCoins`, and changed it so it is not limited to standing limit orders but any active orders even those with epoch status (although that doesn't make a lot of sense on startup when the epoch is empty).  The details of shutdown and start up will need to be considered more carefully in the future before this detail is resolved.

chappjc commented on server/coinlock/coinlocker.go:
```diff
@@ -0,0 +1,237 @@
+
+var _ (CoinLocker) = (*swapLocker)(nil)
+
+type coinIDKey string
```
This is kinda icky, so I'm open to other ideas for a map key, although since there's no restriction on CoinID length in bytes, I'm at a loss for ideas.

buck54321 replied:
I'm okay with strings. I know they take a performance hit in map look ups, but I'm guessing it's not significant enough to warrant the complexity of extra IDs.

chappjc commented on server/db/driver/pg/orders.go:
```diff
@@ -57,7 +58,13 @@ func (coins *dbCoins) Scan(src interface{}) error {
+		bc := make([]byte, cLen)
+		copy(bc, b[1:cLen+1])
+		c = append(c, bc)
+
```
This was subtle and hard to catch, but the lib/pq driver did hit it and the backing buffers were reused.

buck54321 commented on server/market/market.go:
```diff
@@ -385,11 +550,14 @@ func (m *Market) processEpoch(epoch *EpochQueue) {
 			epoch.Epoch, epoch.Duration)
+		// Since Swapper.Negotiate is called asynchronously, we must lock coins
+		// with the Swapper first. The swapper will unlock the coins.
+		m.swapper.LockOrdersCoins(swapOrders)
```
I'm not sure I follow. Why can't the Swapper lock the coins asynchronously?

chappjc replied:
I believe what could happen is a new order could arrive that is backed by coins that should be locked by the swapper, but since execution of `go m.swapper.Negotiate` is at the mercy by the Go scheduler, the new order may get accepted if the locking happens inside `Negotiate`.
